adapters/socketio_mbed_os5.c

```diff
@@ -0,0 +1,495 @@
+        {
+            result = __FAILURE__;
+        }
+        else
+        {
+            if (tcpsocketconnection_connect(socket_io_instance->tcp_socket_connection, socket_io_instance->hostname, socket_io_instance->port) != 0)
+            {
+                tcpsocketconnection_destroy(socket_io_instance->tcp_socket_connection);
+                socket_io_instance->tcp_socket_connection = NULL;
+                result = __FAILURE__;
+            }
+            else
+            {
+                tcpsocketconnection_set_blocking(socket_io_instance->tcp_socket_connection, false, 0);
+
+                socket_io_instance->on_bytes_received = on_bytes_received;
+                socket_io_instance->on_bytes_received_context = on_bytes_received_context;
+
+                socket_io_instance->on_io_error = on_io_error;
+                socket_io_instance->on_io_error_context = on_io_error_context;
+
+                socket_io_instance->io_state = IO_STATE_OPEN;
+
+                result = 0;
+            }
+        }
+    }
+    
+    if (on_io_open_complete != NULL)
+    {
+        on_io_open_complete(on_io_open_complete_context, result == 0 ? IO_OPEN_OK : IO_OPEN_ERROR);
+    }
+
+    return result;
+}
+
+int socketio_close(CONCRETE_IO_HANDLE socket_io, ON_IO_CLOSE_COMPLETE on_io_close_complete, void* callback_context)
+{
+    int result = 0;
+
+    if (socket_io == NULL)
+    {
+        result = __FAILURE__;
+    }
+    else
+    {
+        SOCKET_IO_INSTANCE* socket_io_instance = (SOCKET_IO_INSTANCE*)socket_io;
+        if (socket_io_instance->io_state == IO_STATE_CLOSED || 
+            socket_io_instance->io_state == IO_STATE_ERROR)
+        {
+            result = __FAILURE__;
+        }
+        else
+        {
+            close_tcp_connection(socket_io_instance);
+            if (on_io_close_complete != NULL)
+            {
+                on_io_close_complete(callback_context);
+            }
+            result = 0;
+        }
+    }
+    
+    return result;
+}
+
+int socketio_send(CONCRETE_IO_HANDLE socket_io, const void* buffer, size_t size, ON_SEND_COMPLETE on_send_complete, void* callback_context)
+{
+    int result;
+
+    if ((socket_io == NULL) ||
+        (buffer == NULL) ||
+        (size == 0))
+    {
+        /* Invalid arguments */
+        result = __FAILURE__;
+    }
+    else
+    {
+        result = 0;
+
+        SOCKET_IO_INSTANCE* socket_io_instance = (SOCKET_IO_INSTANCE*)socket_io;
+        if (socket_io_instance->io_state != IO_STATE_OPEN)
+        {
+            result = __FAILURE__;
+        }
+        else
+        {
+            // Queue the data, and the socketio_dowork sends the package later
+            if (add_pending_io(socket_io_instance, buffer, size, on_send_complete, callback_context) != 0)
+            {
+                result = __FAILURE__;
+            }
+            
+            if (send_queued_data(socket_io_instance) < 0)
+            {
+                result = __FAILURE__;
+            }
```

Comment:
We shouldn't send queued data here. Should just be in dowork, otherwise we have a race condition

Reply:
good catch, fixed